recirq/quantum_chess/swap_updater_test.py

```diff
@@ -0,0 +1,117 @@
+    assert list(updater.update_iteration()) == [cirq.CNOT(Q[1], Q[2])]
+    assert list(updater.update_iteration()) == [cirq.CNOT(Q[5], Q[2])]
+    # Next iteration adds a swap between Q1 and Q4.
+    assert list(updater.update_iteration()) == [cirq.SWAP(Q[1], Q[4])]
+    # Remaining gates are added as-is.
+    assert list(updater.update_iteration()) == [cirq.CNOT(Q[4], Q[5])]
+    assert list(updater.update_iteration()) == [cirq.CNOT(Q[1], Q[4])]
+    assert list(updater.update_iteration()) == [cirq.CNOT(Q[4], Q[3])]
+    # The final two gates are added in the same iteration, since they operate on
+    # mutually exclusive qubits and are both simultaneously active.
+    assert set(updater.update_iteration()) == {
+        cirq.CNOT(Q[5], Q[4]), cirq.CNOT(Q[3], Q[0])
+    }
+
+
+def test_example_9():
+    Q = FIGURE_9A_PHYSICAL_QUBITS
+    initial_mapping = dict(zip(q, Q))
+    updater = SwapUpdater(FIGURE_9A_CIRCUIT, Q, initial_mapping,
+                          lambda q1, q2: [cirq.SWAP(q1, q2)])
+    updated_circuit = cirq.Circuit(
+        SwapUpdater(FIGURE_9A_CIRCUIT, Q, initial_mapping,
+                    lambda q1, q2: [cirq.SWAP(q1, q2)]).add_swaps())
+    assert updated_circuit == cirq.Circuit(cirq.SWAP(Q[0], Q[1]),
+                                           cirq.CNOT(Q[1], Q[2]),
+                                           cirq.CNOT(Q[5], Q[2]),
+                                           cirq.SWAP(Q[1], Q[4]),
+                                           cirq.CNOT(Q[4], Q[5]),
+                                           cirq.CNOT(Q[1], Q[4]),
+                                           cirq.CNOT(Q[4], Q[3]),
+                                           cirq.CNOT(Q[5], Q[4]),
+                                           cirq.CNOT(Q[3], Q[0]))
+
+
+def test_pentagonal_split_and_merge():
+    grid_3x2 = list(
+        cirq.GridQubit(row, col) for row in range(2) for col in range(3))
+    logical_qubits = list(
+        cirq.NamedQubit(f'{x}{i}') for x in ('a', 'b') for i in range(3))
+    initial_mapping = dict(zip(logical_qubits, grid_3x2))
+    a1, a2, a3, b1, b2, b3 = logical_qubits
+    circuit = cirq.Circuit(qm.normal_move(a1, b1), qm.normal_move(a2, a3),
+                           qm.merge_move(a3, b1, b3))
+
+    updated_circuit = cirq.Circuit(
+        SwapUpdater(circuit, grid_3x2, initial_mapping).add_swaps())
+    for op in updated_circuit.all_operations():
+        assert len(op.qubits) == 2
+        q1, q2 = op.qubits
+        assert q1 in grid_3x2
+        assert q2 in grid_3x2
+        assert q1.is_adjacent(q2)
+
+
+def test_decomposed_swaps():
+    Q = FIGURE_9A_PHYSICAL_QUBITS
+    initial_mapping = dict(zip(q, Q))
+    updater = SwapUpdater(FIGURE_9A_CIRCUIT, Q, initial_mapping,
+                          generate_decomposed_swap)
+    # First iteration adds a swap between Q0 and Q1.
+    # generate_decomposed_swap implements that swap operation as four
+    # sqrt-iswaps.
```

Comment:
Instead of testing the implementation, we should test the effect.  For instance, that the effect is equivalent to swapping Q[0] and Q[1] (via cirq.unitary or some other method).

Reply:
great idea.

Related question: how is the unitary matrix index determined for a circuit's qubits? For some of these tests, it would be cool to compare the logical circuit's unitary before and after adding swaps; they should be the same, but only after ignoring any ancillary qubits.